doc/nrf/libraries/networking/nrf_provisioning.rst

```diff
@@ -0,0 +1,162 @@
+  In this case it is assumed that a network connection has been established in the same phase.
+
+The function uses the following two arguments:
+
+*  A pointer to a callback function which is to be called on modem state changes.
+*  A pointer that must be called once provisioning is done.
+
+If you provide ``null`` as a callback function address argument, a corresponding default callback used.
+Consequent calls to the initialization function will only change the callback functions.
+This behavior is beneficial when the client has been initialized during the system initialization phase, but the application wants to register its own callback functions afterwards.
+
+Provisioning
+============
+
+By default, when provisioning is done after receiving the ``FINISHED`` command, the device is rebooted.
+The behavior can be overwritten by providing a unique callback function for the initialization function.
+
+If anything is written on the modem's non-volatile memory, the modem needs to be set in offline mode.
+This is because the modem cannot be connected when any data is to be written on to its storage area.
+After the memory write is completed, the aforementioned callback function must be called again to set the modem in desired state.
+To use the default implementation, ``NULL`` can be passed as an argument to the :c:func:`nrf_provisioning_init` function.
+The default callback function must be copied and modified based on your needs if it is necessary.
+
+The library starts provisioning when it initializes and afterwards it is done in a configured interval.
+The interval is read from the storage settings, and it can be updated with a provisioning command like any other key-value pair.
+
+During provisioning, the library first tries to establish the transport for communicating with the service.
+This procedure involves a TLS handshake where the client establishes the correct server.
+The server uses the attestation token or the JWT token generated by the device for authenticating the client.
+See :ref:`lib_modem_attest_token` and :ref:`lib_modem_jwt` for more information on client authentication.
+
+The TLS handshake needs to be done twice:
+
+* Before requesting commands
+* After the execution of the commands to report the results.
+
+The library shuts down the modem for writing data to the modem's non-volatile memory, if `AT commands <AT Commands Reference Guide_>`_ are involved.
+Once the memory writes are complete, connection is re-established to report the results back to the server.
+The results are reported back to the server when either all the commands succeed or when an error occurs.
+If an error occurs, the results of all the commands that are successfully executed before the error and the erroneous result are reported back to the server.
+All successfully executed commands will be removed from the server-side queue, but if any errors occur, the erroneous command and all the remaining unexecuted commands will be flushed from the queue.
```

Comment:
We need to be careful to distinguish this from the provisioning configuration for a claimed device in nRF Cloud: these are, as I understand, the commands we're talking about here.

On the Cloud side, the commands form the *provisioining configuration*, and the configuration remains in the interface and the service's data about the device until the commands are deleted by the user. 

If this is some queue in the network, I'd like that to be a little clearer.

Reply:
Yes, we are talking about the provisioning configuration here.

Reply:
> All successfully executed commands will be removed from the server-side queue, but if any errors occur, the erroneous command and all the remaining unexecuted commands will be flushed from the queue.

Just to be clear whether you are describing server functionality or device functionality, the removal from the queue and flushing of the queue are done on the server side, not on the device side in this library, correct?